src/compiler/checker.ts

```diff
@@ -17967,7 +17967,10 @@ namespace ts {
 
             for (let i = 0; i < paramCount; i++) {
                 const sourceType = i === restIndex ? getRestTypeAtPosition(source, i) : tryGetTypeAtPosition(source, i);
-                const targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+                let targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+                if (i === restIndex && targetType && sourceType && isTupleType(sourceType) && !sourceType.target.hasRestElement) {
+                    targetType = mapType(targetType, t => isTupleType(t) && !t.target.hasRestElement ? sliceTupleType(t, 0, t.target.fixedLength - sourceType.target.fixedLength) : t);

```

Comment:
Ideally, I should drop `!t.target.hasRestElement` from here because I think that rest in the target tuple could be handled. For instance this should typecheck OK:
```ts
// Type '(a: string | number, b: string | number) => void' is not assignable to type '(...args: [number, string, ...boolean[]] | [string, number, ...boolean[]]) => void'.
//   Types of parameters 'a' and 'args' are incompatible.
//     Type '[number, string, ...boolean[]] | [string, number, ...boolean[]]' is not assignable to type '[a: string | number, b: string | number]'.
//       Type '[number, string, ...boolean[]]' is not assignable to type '[a: string | number, b: string | number]'.
//         Target allows only 2 element(s) but source may have more.(2322)
const f: (
  ...args: [number, string, ...boolean[]] | [string, number, ...boolean[]]
) => void = (a, b) => {};
```

This would require a different kind of a tuple slice - or rather, a tuple expansion (?). I think that the correct answer would be to convert a tuple type with rest to a tuple type of completely fixed elements (since we are trying to match a fixed-size source). 

I've looked throughout the codebase for an algorithm like that and I couldn't find it - perhaps you would know if something like this is already implemented anywhere? If this isn't implemented anywhere then I can probably write it, I've just figured out that it might be best to first ask about it cause it feels that it might already exist somewhere.

Reply:
It feels like this should exist, given one can write:

```ts
type AtLeastOne = [number, ...number[]];
type AtLeastTwo = [number, number, ...number[]];
type AtLeastThree = [number, number, number, ...number[]];

declare let one: AtLeastOne;
declare let two: AtLeastTwo;
declare let three: AtLeastThree;
declare let twoOrThree: AtLeastTwo | AtLeastThree;

one = two;
one = three;
one = twoOrThree;
```

In that the function can accept fewer parameters and it's legal, just like assigning a "bigger" tuple is to a smaller tuple that optionally takes more.

The "Target allows only 2 element(s) but source may have more.(2322)" message is the bug that this PR is trying to fix, but the fact that it only works in one way (if I am understanding this correctly) leaves me a little suspicious that there might be something missing.

Reply:
Thanks for the given snippet! I will analyze this and compare how the compiler checks those. I'm a little bit busy right now but I hope to get back to this PR within 2 weeks.

Reply:
I'm slowly looking back at this since I'm also trying to fix a different, related, issue. In your examples, all those types have a variadic length whereas all (or most) of the situations that I already handle here and which I was thinking about here include a fixed size tuple,

Weirdly enough I have just discovered that this is not a valid assignment today:
```ts
declare let atLeastOne: [number, ...number[]];

const justOne: [number] = atLeastOne
// Type '[number, ...number[]]' is not assignable to type '[number]'.(2322)
```

I don't understand why this limitation is in place here as intuitively I would consider a type with *more* keys to be always assignable to one with *less* keys. After all, this works perfectly fine:
```ts
const foo: { a: 1 } = {} as { a: 1, b: 2 }
```

Reply:
I guess the reason why we can't assign to fixed size tuples like this is their `length` property that is typed as a discrete number.